### PR TITLE
Fix arithmetic shift

### PR DESCRIPTION
**Pull request by alexcarsello**
Keyi and I discovered a bug in the pe shifter today. The shift operands were not declared as signed nets in test_shifter.svp, so even though the arithmetic shift operator (>>>) is used, the sign extension does not occur. I've fixed that with this PR.

I also set the default memory tile height to 1 in top.vp. I'm not sure how this got reverted...

### CODE REVIEW

**Comment by leonardt** [comment]
Unfortunate this slipped through, the issue is that ashr is not defined in the spec we were working off of (https://docs.google.com/document/d/1OXBBVk1exV1Fa65yBVIGxb6pd33UTEOqxWl0fBEhGRU/edit#heading=h.wbpe5nq71b6a), so the unit tests do not cover this operation.

**Comment by tema8** [comment]
To get arithmetic shift, you need to set 'Signed' field (bit 6) in op_code

**Comment by alexcarsello** [comment]
Yes, but it doesn't work because the operands are not declared as signed in the verilog. If the operands are not signed, >>> is the same as >>, so it shifts in zeros regardless of whether or not the 'signed' field of the op_code is set.

**Comment by steveri** [comment]
I will approve this once the checks stop failing...I am looking into this now...

**Comment by leonardt** [comment]
The pe spec implementation needs to be updated to handle the signed bit, the current tests assume the shr is always logical. Here's the change for the spec https://github.com/phanrahan/pe/commit/c34ef5e49a037f50f2b107662504e38b9ccb5586
pushing a change to this branch to use that pe branch.

**Comment by steveri** [comment]
Okay, this is very weird.
https://travis-ci.org/StanfordAHA/CGRAGenerator/builds/421809752
It looks like lenny's test is failing on a *left shift* when both operands are negative.  it looks like it is right-filling the shift with zeroes!!??  E.g. down around line 4781:

```
[Test Iteration 128] Inputs: data0=fff9, data1=fff8, bit0=0, bit1=0, bit2=0
    expected_res=f900, actual_res=f9ff
```
Lenny, I just saw your new comment, but will that fix this??

**Comment by leonardt** [comment]
It looks like this change actually causes the lshl tests to fail, it seems like it's doing some sort of signed shift for left shift as well?

**Comment by leonardt** [comment]
My change doesn't affect the left shift operator. The expected value looks like it's shifting left by 8 places, appending zeros, the actual value looks like it's shifting left by 8 places and appending ones.

File: hardware/generator_z/top/top.vp [hunk 1]
```diff
@@ -11,7 +11,7 @@
 //; my $mem_tile_height = parameter(
 //;    doc => 'height of memtile relative to minimum grid height',
 //;    name=>'mem_tile_height',
-//;    val => 2,
+//;    val => 1,
```
Comment (leonardt) [inline hardware/generator_z/top/top.vp]: Is this change related to shift?

Comment (alexcarsello) [reply hardware/generator_z/top/top.vp]: No, it's not. Sorry I should've made a separate PR for that

**Comment by alexcarsello** [comment]
Yes, as long as it still performs the signed arithmetic shift properly, which I think it does.

**Comment by leonardt** [comment]
Here's the failure for signed right shift in the compute unit
```
[Test Iteration 0] Inputs: op_a=fff8, op_b=fff8, op_d_p=0
    expected_res=ffff, actual_res=ff
```
It should be shifting by the lower 4 bits of op_b, so by 8. The expected result looks good, shifting by 8 and sign extending. The actual result looks off, not exactly sure what the result implies.

One question, the `pe.py` does left and right shift by using `b[:4]` (the lower 4 bits), is this expected behavior (that is, is this the spec)? Or should it be using all the bits

**Comment by steveri** [comment]
This is the left shift code (below).  It looks like, instead of doing left shift it does bitreverse+rightshift+bitreverse with obvious bad consequence for signed operands (i.e LSB-fill).
Anybody know why it does it this way??

```
assign shift_inp = dir_left ? a_reverse : a;
assign shift_res =  is_signed ? (shift_inp >>> b): (shift_inp >> b);//(shift_inp >> b);

assign rshift_val = shift_res;//(a >> b);

genvar g_i;
generate
  for(g_i = 0; g_i < DataWidth; g_i++) begin: GEN_L_SHIFT
    assign a_reverse[g_i]  = a[DataWidth-1-g_i];
    assign lshift_val[g_i] = shift_res[DataWidth-1-g_i];
  end
endgenerate

assign res = dir_left ? lshift_val: rshift_val;

```

**Comment by alexcarsello** [comment]
Lenny, the reason I made b signed is because I think all of the operands must be signed in order for it to perform a signed shift. It looks like making b unsigned brought us back to where we were before. I think we have 2 options here:

1. Change the spec to make the shift amount signed, rather than unsigned.

2. Leave the spec as is, and in the RTL, make b signed, and shift by |b|, which we can assign to another signed net.

**Comment by alexcarsello** [comment]
Wait, never mind. That doesn't make sense. I think we may just have to do the sign extension ourselves.

**Comment by leonardt** [comment]
Can someone tell me what the expected result *should* be for shift right in signed mode for these operands? Is `0xffff` correct? (shifting by the lower 4 bits of op_b)
```
[Test Iteration 0] Inputs: op_a=fff8, op_b=fff8, op_d_p=0
    expected_res=ffff, actual_res=ff
```

**Comment by alexcarsello** [comment]
yeah, I just ran my own tests, and found that, too. 0xffff is correct. I'm not sure why it's not performing the sign extension on that example... it is working properly in my tests.

**Comment by alexcarsello** [comment]
Looks like this is good to go... any objections?

File: .travis.yml [hunk 2]
```diff
@@ -52,7 +52,7 @@ script:
 
   # Install python dependencies
   - pip install pytest numpy delegator.py bit_vector
-  - pip install git+git://github.com/phanrahan/pe.git
+  - pip install git+git://github.com/phanrahan/pe.git@signed-rshift
```
Comment (leonardt) [inline .travis.yml]: When this eventually gets into master, we should change this back to installing from the `pe` master branch.
